src/ssl.jl

```diff
@@ -373,12 +387,12 @@ end
 """
     SSLStream.
 """
-mutable struct SSLStream <: IO
+mutable struct SSLStream{T} <: IO
     ssl::SSL
     ssl_context::SSLContext
     rbio::BIO
     wbio::BIO
-    io::TCPSocket
+    io::T
```

Comment:
Idea: If we parametrize `BIO` on the `io` type it was created with we can have more type stable code downstream and save ourselves some unrolling on the result of `bio_get_data`. Right now, we'd allocate a bunch for, e.g., a `UDPSocket` due to dynamic dispatch. The default case when we don't know the IO type for whatever reason can use `Any` or abstract `IO`.

I tried to prototype this idea but I hit a problem with how callbacks are used. To get any benefit from this new parametrization, we'd need to call the corresponding specialized method (obviously), but now we use a single method globally via `BIOStreamCallbacks`, e.g. `@cfunction on_bio_stream_read Cint (BIO, Ptr{Cchar}, Cint)` but we'd need 
```
@cfunction on_bio_stream_read Cint (BIO{TCPSocket}, Ptr{Cchar}, Cint)
@cfunction on_bio_stream_read Cint (BIO{UDPSocket}, Ptr{Cchar}, Cint)
...
``` 
to be used by the `BIOMethod` object accordingly... which would also mean somehow correlating the type parameter with the `BIOMethod` and then maintaining keeping the `@cfunctions` around. I'm not too sure how to best approach that, so I'll just leave this here in case you have some better idea:)

Reply:
https://github.com/JuliaWeb/OpenSSL.jl/pull/27

There are two BIOMethods, one specialize for IO another for TCPStream.